compareBuckets/utils.js

```diff
@@ -0,0 +1,106 @@
+const http = require('http');
+const async = require('async');
+const { URL } = require('url');
+
+const { jsutil } = require('arsenal');
+
+const httpAgent = new http.Agent({
+    keepAlive: true,
+});
+
+// TODO: remove duplication
+function httpRequest(method, url, cb) {
+    const cbOnce = jsutil.once(cb);
+    const urlObj = new URL(url);
+    const req = http.request({
+        hostname: urlObj.hostname,
+        port: urlObj.port,
+        path: `${urlObj.pathname}${urlObj.search}`,
+        method,
+        agent: httpAgent,
+    }, res => {
+        if (method === 'HEAD') {
+            return cbOnce(null, res);
+        }
+        const chunks = [];
+        res.on('data', chunk => chunks.push(chunk));
+        res.once('end', () => {
+            const body = chunks.join('');
+            // eslint-disable-next-line no-param-reassign
+            res.body = body;
+            return cbOnce(null, res);
+        });
+        res.once('error', err => cbOnce(new Error(
+            'error reading response from HTTP request '
+                + `to ${url}: ${err.message}`
+        )));
+        return undefined;
+    });
+    req.once('error', err => cbOnce(new Error(
+        `error sending HTTP request to ${url}: ${err.message}`
+    )));
+    req.end();
+}
+
+function getObjectURL(bucket, objectKey) {
+    if (!bucket) {
+        return 's3://';
+    }
+    if (!objectKey) {
+        return `s3://${bucket}`;
+    }
+    return `s3://${bucket}/${encodeURI(objectKey)}`;
+}
+
+function getBucketdURL(hostPort, params) {
+    let baseURL = `http://${hostPort}/default/bucket/${params.Bucket}`;
+    if (params.Key !== undefined) {
+        baseURL += `/${encodeURIComponent(params.Key)}`;
+    }
+    const queryParams = [];
+    if (params.MaxKeys) {
+        queryParams.push(`maxKeys=${params.MaxKeys}`);
+    }
+    if (params.KeyMarker !== undefined) {
+        queryParams.push(`marker=${encodeURIComponent(params.KeyMarker)}`);
+    }
+    return `${baseURL}${queryParams.length > 0 ? '?' : ''}${queryParams.join('&')}`;
+}
+
+function listBucketMasterKeys(params, cb) {
```

Comment:
We need to wrap this function with a retry loop, otherwise the script will stop at the first error, which is likely to happen on a large bucket

Reply:
adding `async.retry` around the `httpRequest`